src/include/txn.i

```diff
@@ -262,36 +259,32 @@ __wt_txn_update_set_timestamp(WT_SESSION_IMPL *session,
 	wt_timestamp_t *timestamp;
 	bool needs_timestamp;
 
-	txn = &session->txn;
-	upd = op->u.op_upd;
 	if (was_set != NULL)
 		*was_set = false;
 
-	/*
-	 * The timestamp is in the page deleted structure for truncates, or
-	 * in the update for other operations.
-	 */
-	if (op->type == WT_TXN_OP_REF_DELETE)
-		timestamp = op->u.ref == NULL || op->u.ref->page_del == NULL ?
-		    NULL : &op->u.ref->page_del->timestamp;
-	else
-		timestamp = op->u.op_upd == NULL ?
-		    NULL : &op->u.op_upd->timestamp;
+	txn = &session->txn;
 
 	/*
 	 * Updates in the metadata never get timestamps (either now or at
 	 * commit): metadata cannot be read at a point in time, only the most
 	 * recently committed data matches files on disk.
 	 */
-	needs_timestamp = !WT_IS_METADATA(op->btree->dhandle) &&
-	    F_ISSET(txn, WT_TXN_HAS_TS_COMMIT) &&
-	    (timestamp == NULL || __wt_timestamp_iszero(timestamp) ||
-	    F_ISSET(txn, WT_TXN_PREPARE));
+	if (WT_IS_METADATA(op->btree->dhandle) ||
+	    !F_ISSET(txn, WT_TXN_HAS_TS_COMMIT))
+		return;
 
+	/*
+	 * The timestamp is in the page deleted structure for truncates, or
+	 * in the update for other operations.
+	 */
+	timestamp = op->type == WT_TXN_OP_REF_DELETE ?
+	    &op->u.ref->page_del->timestamp : &op->u.op_upd->timestamp;
```

Comment:
This simplification seems OK to me right now, but we won't always have a valid `op->u.op_upd` - that's part of the ongoing prepared transaction work. I don't like nested ternary statements, which is why I exploded it out.
ps: I know I can't expect you to know what I'm thinking will happen soon.

Reply:
> This simplification seems OK to me right now, but we won't always have a valid `op->u.op_upd`

Then maybe the clearest choice is to explicitly handle a NULL timestamp reference now.